### ISSUE TITLE
Fine-tuning model using knowledge base

### ISSUE THREAD

**Issue by pokhiii** [ColoredCow/ai-voice-assistant#3]
> _Note: We are using the [`model meta-llama/Llama-3.2-1B-Instruct`](https://github.com/ColoredCow/ai-voice-assistant/blob/0626048a4a2ae93de1517c91dfab10d4e426f3cc/server.py#L17) for generating responses._

## Step 1: Set Up Environment for Fine-Tuning

We’ll need

1. Python and PyTorch
2. Hugging Face’s transformers library to work with the model and datasets to handle the dataset
3. A GPU or a cloud service (like AWS or Google Colab) for faster processing, as fine-tuning can be compute-intensive

## Step 2: Preparing Dataset

1. Format data as a list of question-answer pairs in a `.json` file
   ```json
   [
       {"question": "What crop is most profitable in Nashik?", "answer": "In Nashik's climate, grapes and pomegranates are highly profitable."},
       {"question": "How can I control pests in rice fields?", "answer": "You can use integrated pest management techniques, including biological controls and safe pesticides."}
   ]
   ```

2. **Load and tokenize** the data in the fine-tuning script.

## Step 3: Load Pre-trained Model and Dataset in the Script

Basic script to load the model, prepare the dataset, and start fine-tuning.
```python
import torch
from transformers import AutoTokenizer, AutoModelForCausalLM, Trainer, TrainingArguments, Dataset

# Load the model and tokenizer
model_name = "meta-llama/Llama-3.2-1B-Instruct"
tokenizer = AutoTokenizer.from_pretrained(model_name)
model = AutoModelForCausalLM.from_pretrained(model_name)

# Load and preprocess the dataset
from datasets import load_dataset

# Assume you have a JSON file with question-answer pairs
dataset = load_dataset('json', data_files='path_to_the_dataset.json')

# Tokenize the data
def preprocess_function(examples):
    inputs = examples['question']
    targets = examples['answer']
    model_inputs = tokenizer(inputs, max_length=128, truncation=True, padding="max_length")
    labels = tokenizer(targets, max_length=128, truncation=True, padding="max_length")['input_ids']
    model_inputs["labels"] = labels
    return model_inputs

tokenized_dataset = dataset.map(preprocess_function, batched=True)

### Step 4: Set Up Training Arguments
training_args = TrainingArguments(
    output_dir="./fine_tuned_model",
    evaluation_strategy="epoch",
    learning_rate=2e-5,
    per_device_train_batch_size=2,
    per_device_eval_batch_size=2,
    num_train_epochs=3,
    weight_decay=0.01,
    logging_dir='./logs'
)

# Define the trainer
trainer = Trainer(
    model=model,
    args=training_args,
    train_dataset=tokenized_dataset["train"],
    eval_dataset=tokenized_dataset["test"],
)

# Start fine-tuning
trainer.train()
```

### Step 5: Fine-Tune the Model

1. **Run the script.** It will load the dataset, tokenize the question-answer pairs, and begin fine-tuning.
2. Monitor the training to ensure it’s progressing well and adjust hyperparameters (like `learning_rate` or `num_train_epochs`) if necessary.

### Step 6: Save and Test the Fine-Tuned Model

```python
model.save_pretrained("fine_tuned_llama")
tokenizer.save_pretrained("fine_tuned_llama")
```

### Step 7: Integrate the Fine-Tuned Model in the App

```python
# Replace with the path to the fine-tuned model
model_path = "fine_tuned_llama"
pipe = pipeline(
    "text-generation",
    model=model_path,
    tokenizer=model_path,
    torch_dtype=torch.bfloat16,
    device_map="auto",
)
```

### Tips for Fine-Tuning

- Start with a small learning rate to avoid large weight changes, which can disrupt the model’s knowledge.
- Use a lower number of training epochs initially (e.g., 3-5) and adjust based on results.
- Fine-tuning requires a good GPU for efficiency; consider using cloud resources like Google Colab or AWS if you don’t have access to one.

### Optional: Adding Contextual Memory for Conversational Flow

For follow-up questions, we should consider adding a retrieval component (like using embeddings to search for relevant past answers) so the bot can refer to previous answers, making it feel more conversational and context-aware.
